File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []
        def backtrack(cur):
            if len(cur) == len(nums):
                ans.append(cur[:])
                return
            
            for num in nums:
                if num not in cur:
                    cur.append(num)
                    backtrack(cur)
                    cur.pop()

        backtrack([])
        return ans

File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> cur = {};
        backtrack(nums,ans,cur);
        return ans;
    }
    void backtrack(vector<int>& nums,vector<vector<int>>& ans,vector<int>& cur){
        if(cur.size()==nums.size()){
            ans.push_back(cur);
            return;
        }
        for(int num:nums){
            if(find(cur.begin(),cur.end(),num) == cur.end()){
                cur.push_back(num);
                backtrack(nums,ans,cur);
                cur.pop_back();
            }
        }
    }
};